File: ios/Classes/EmmaSerializer.swift
//
//  EmmaSerializer.swift
//  
//
//  Created by Adrián Carrera on 12/11/2020.
//

import EMMA_iOS

class EmmaSerializer {
    static func nativeAdToDictionary(_ nativeAd: EMMANativeAd) -> [String: Any?] {
        return [
            "id": nativeAd.idPromo,
            "templateId": nativeAd.nativeAdTemplateId ?? "",
            "cta": nativeAd.getField("CTA"),
            "times": nativeAd.times,
            "tag": nativeAd.tag ?? NSNull(),
            "params": nativeAd.params ?? [:],
            "showOn": nativeAd.openInSafari ? "browser" : "inapp",
            "fields": nativeAd.nativeAdContent as? [String: Any] ?? []
            ]
    }
    
    static func inAppTypeFromString(inAppType: String) -> InAppType? {
        switch inAppType {
            case "startview":
                return .Startview
            case "nativeAd":
                return .NativeAd
            default:
                return nil
        }
    }
    
    static func inAppTypeToCommType(type: InAppType) -> EMMACampaignType? {
        switch type {
        case InAppType.Startview:
            return EMMACampaignType.campaignStartView
        case InAppType.NativeAd:
            return EMMACampaignType.campaignNativeAd
        default:
            return nil
        }
    }
}
            

File: android/src/main/kotlin/io/emma/emma_flutter_sdk/EmmaSerializer.kt
package io.emma.emma_flutter_sdk

import io.emma.android.enums.CommunicationTypes
import io.emma.android.model.EMMACampaign
import io.emma.android.model.EMMANativeAd
import io.emma.android.model.EMMANativeAdField
import io.emma.android.utils.EMMALog


object EmmaSerializer {

    fun nativeAdToMap(nativeAd: EMMANativeAd): Map<String, Any>? {
        val nativeAdMap = HashMap<String, Any>()
        try {
            nativeAdMap["id"] = nativeAd.campaignID.toInt()
            nativeAdMap["templateId"] = nativeAd.templateId
            nativeAdMap["times"] = nativeAd.times.toInt()
            nativeAdMap["tag"] = nativeAd.tag
            nativeAdMap["cta"] = nativeAd.campaignUrl
            nativeAdMap["showOn"] = if(nativeAd.showOnWebView()) "inapp" else "browser"
            nativeAdMap["params"] = nativeAd.params
            nativeAdMap["fields"] = nativeAdFieldsToMap(nativeAd.nativeAdContent)
        } catch (e: Exception) {
            EMMALog.e("Error parsing native ad", e)
            return null
        }
        return nativeAdMap
    }

    private fun nativeAdFieldsToMap(fields: Map<String, EMMANativeAdField>): Map<String, String> {
        val fieldsMap = HashMap<String, String>()
        for ((_, value) in fields.entries) {
            fieldsMap[value.fieldName] = value.fieldValue
        }
        return fieldsMap
    }

    fun getInAppRequestTypeFromString(type: String): EMMACampaign.Type? {
        when(type) {
            "startview" -> {
                return EMMACampaign.Type.STARTVIEW
            }
            "nativeAd" -> {
                return EMMACampaign.Type.NATIVEAD
            }
            else -> {
                return null
            }
        }
    }

    fun inAppTypeToCommType(type: EMMACampaign.Type?): CommunicationTypes? {
        if (type == null) {
            return null
        }

        when(type) {
            EMMACampaign.Type.STARTVIEW -> {
                return CommunicationTypes.STARTVIEW
            }
            EMMACampaign.Type.NATIVEAD -> {
                return CommunicationTypes.NATIVE_AD
            }
            else -> {
                return null
            }
        }
    }

}